### ISSUE TITLE
Wrap and TileMovement Issue

### ISSUE THREAD

**Issue by Learz** [Scirra/Construct-3-bugs#2491]
## Problem description

Using the wrap behavior with the TileMovement behavior makes the object Target out of bound and therefore makes the object move infinitely in that direction since it cannot reach it's target. I believe the target should be set to the other side of the scene and not a negative value.

Edit : A similar issue happens when using the BoundToLayout Behavior, the sprite gets stuck on the side, unable to move after touching it.

## Attach a .c3p

[WrapTileIssue.zip](https://github.com/Scirra/Construct-3-bugs/files/2731096/WrapTileIssue.zip)

## Steps to reproduce

1. Press the left key on your keyboard until the sprite wraps around the scene.

## Observed result

The sprite will keep moving in the same direction forever.
The target tile is displayed on the top left of the screen, displaying a value that will never be reached.

## Expected result

The sprite should stop moving after arriving at the desired tile from other side of the scene.
The target tile should be the value from the other side of the scene and not out of the scene.

## First affected release

r131 stable

## Affected browsers/platforms

* Chrome: yes
* Firefox: yes
* Edge: yes
* Safari: yes

## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 71.0.3578.98
Browser engine: Blink
Browser architecture: 64-bit
Context: browser
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36
C3 release: r131 (stable)
Language setting: en-US

Local storage
Storage quota (approx): 137 gb
Storage usage (approx): 52 mb (0%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (NVIDIA GeForce GTX 970 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by shortercode** [comment]
Thanks for the report. The tile movement doesn't expect to be put in a position that it cannot reach it's target location, which the wrap behaviour is preventing. This is an oversight but I'm not sure how to resolve this at the moment.

**Comment by Learz** [comment]
I don't know the specifics of the code behind the behavior and the interactability between them, but I believe adding a modulus equal to the scene dimensions to the calculation of both Targets should make it wrap around.

For example, if the scene is 128 pixels wide, and you try to put the X target at -32, adding a %128 would automatically make the value equal 96, which is exactly where we want it to be. The same would work if the value is bigger than 128. For example if the X target equals 160, a %128 would automatically wrap around the value to 32.

I understand if it's not as easy of a solution as I might think, but I hope this helps.

**Comment by AshleyScirra** [comment]
The behaviors work entirely independently and don't necessary know the other exists, which makes this trickier. It would be trivial to write the overall behavior individually, but the interop between separately coded behaviors is sometimes a headache.

**Comment by shortercode** [comment]
I've given this a fair amount of thought, and I've tried a couple of ideas out, but unfortunately I think I'm  going to have to close this as won't fix.

As @AshleyScirra said the major issue here is that the 2 behaviours are ignorant of the others existence.

When you reach the boundary and tell TileMovement to go past it the following happens:
1. It moves(instantly) into the next tile
2. It checks for a collision
3. It moves(instantly) back
4. If there was no collision it updates it's target location

After it has done this it will attempt to move towards this location until it has succeeded.

Both Wrap and BoundToLayout basically just check if the object is in the correct place every frame, and teleport to the correct location if it isn't. As this is not a boundary that the TileMovement understands it cannot do anything special about it. It's possible for it to detect that it has been moved by something other than itself, but if the layout size isn't grid aligned then it can't smoothly transition anyway ( gets stuck on some edges or up "touching" the layout but being completely off screen ).
